### PR TITLE
the default language added to the setting service

### PR DESCRIPTION
**Pull request by 2403905**
## Related Issue

- Fixes [https://github.com/owncloud/enterprise/issues/5915](https://github.com/owncloud/enterprise/issues/5915)

## How Has This Been Tested?

- test environment: Local.
- Please fist read the notice in a readme!
- test case 1: Set SETTINGS_DEFAULT_LANGUAGE=de, run ocis. Log in as admin, the UI language has to be DE. admin creates a file and shares it with einstein. 
Expected result: 
- If einstein did change a language in a setting to a language different from DE then einstein would receive a letter in DE.
- If einstein already changed a language in a setting to language then einstein would receive a letter in selected language. But not all languages that we can see in a setting already exist for email notifications! If the language exits for UI but not exist for the email notifications then email notifications fail to EN.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [X] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: services/settings/pkg/config/config.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ type Config struct {
 
 	ServiceAccountIDAdmin string `yaml:"service_account_id_admin" env:"OCIS_SERVICE_ACCOUNT_ID;SETTINGS_SERVICE_ACCOUNT_ID_ADMIN" desc:"The ID of the service account having the admin role. See the 'auth-service' service description for more details."`
 
+	DefaultLanguage string `yaml:"default_language" env:"SETTINGS_DEFAULT_LANGUAGE" desc:"The default language. If not defined, the English will be used as default."`
```
Comment (mmattel) [inline services/settings/pkg/config/config.go:42]:
```suggestion
	DefaultLanguage string `yaml:"default_language" env:"SETTINGS_DEFAULT_LANGUAGE" desc:"The default language. If not defined, English will be used as default."`
```
Do we work with language codes, or what to enter?

File: services/settings/README.md [hunk 2]
```diff
@@ -71,3 +71,7 @@ Services can set or query ocis *setting values* of a user from settings bundles.
 ## Service Accounts
 
 The settings service needs to know the ID's of service accounts but it doesn't need their secrets. Currently only one service account can be configured which has the admin role. This can be set with the `SETTINGS_SERVICE_ACCOUNT_ID_ADMIN` envvar, but it will also pick up the global `OCIS_SERVICE_ACCOUNT_ID` envvar. Also see the 'auth-service' service description for additional details.
+
+## Default language
+
+The default language can be defined via SETTINGS_DEFAULT_LANGUAGE environment variable. If this variable is not defined, the English language will be used as default.
```
Comment (mmattel) [inline services/settings/README.md:77]:
```suggestion
The default language can be defined via SETTINGS_DEFAULT_LANGUAGE environment variable. If this variable is not defined, English will be used as default.
```

**Comment by mmattel** [comment]
What are the values that can be entered for a language and how to identify which languages are available?

**Comment by 2403905** [comment]
> What are the values that can be entered for a language and how to identify which languages are available?

Good catch. I'll add

File: services/settings/pkg/service/v0/servicedecorator.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	"github.com/owncloud/ocis/v2/services/settings/pkg/config"
+	"github.com/owncloud/ocis/v2/services/settings/pkg/store/defaults"
+)
+
+// NewDefaultLanguageService returns a default language decorator for ServiceHandler.
+func NewDefaultLanguageService(cfg *config.Config, serviceHandler settings.ServiceHandler) settings.ServiceHandler {
+	defaultLanguage := cfg.DefaultLanguage
+	if defaultLanguage == "" {
+		defaultLanguage = "en"
```
Comment (kobergj) [inline services/settings/pkg/service/v0/servicedecorator.go:18]: can we extract `"en"` to a global `var _defaultLanguage`?

File: services/settings/pkg/service/v0/servicedecorator.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+// GetValueByUniqueIdentifiers implements the ValueService interface
+func (s *defaultLanguageDecorator) GetValueByUniqueIdentifiers(ctx context.Context, req *settingssvc.GetValueByUniqueIdentifiersRequest, res *settingssvc.GetValueResponse) error {
+	err := s.ServiceHandler.GetValueByUniqueIdentifiers(ctx, req, res)
+	if err != nil {
+		return err
+	}
+	if res.Value == nil {
+		res.Value = s.withDefaultLanguageSetting(req.AccountUuid)
+	}
```
Comment (kobergj) [inline services/settings/pkg/service/v0/servicedecorator.go:36]: Shouldn't we only add the language if it was actually requested? I mean `GetValueByUniqueIdentifier` sounds to me like I can also get other settings except language here. But if request another value and it is not found, I would still get the default language setting...

File: services/settings/pkg/service/v0/servicedecorator.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+func (s *defaultLanguageDecorator) withDefaultLanguageSetting(accountUUID string) *settingsmsg.ValueWithIdentifier {
+	bundle := generateBundleProfileRequest()
+	return &settingsmsg.ValueWithIdentifier{
+		Identifier: &settingsmsg.Identifier{
+			Extension: bundle.Extension,
+			Bundle:    bundle.Name,
+			Setting:   "language",
+		},
+		Value: &settingsmsg.Value{
+			Id:          "",
```
Comment (kobergj) [inline services/settings/pkg/service/v0/servicedecorator.go:65]: Doesn't it need an ID?

**Review by kobergj — CHANGES_REQUESTED** [review]

File: services/settings/README.md [hunk 6]
```diff
@@ -71,3 +71,7 @@ Services can set or query ocis *setting values* of a user from settings bundles.
 ## Service Accounts
 
 The settings service needs to know the ID's of service accounts but it doesn't need their secrets. Currently only one service account can be configured which has the admin role. This can be set with the `SETTINGS_SERVICE_ACCOUNT_ID_ADMIN` envvar, but it will also pick up the global `OCIS_SERVICE_ACCOUNT_ID` envvar. Also see the 'auth-service' service description for additional details.
+
+## Default language
+
+The default language can be defined via SETTINGS_DEFAULT_LANGUAGE environment variable. If this variable is not defined, English will be used as default. The value has ISO 639-1 format ("de", "en", etc.) and limited by the supported languages. This setting can be used to set the default language for invitation emails.
```
Comment (mmattel) [inline services/settings/README.md:77]:
```suggestion
The default language can be defined via SETTINGS_DEFAULT_LANGUAGE environment variable. If this variable is not defined, English will be used as default. The value has the ISO 639-1 format ("de", "en", etc.) and is limited by the list supported languages. This setting can be used to set the default language for invitation emails.
```

**Comment by mmattel** [comment]
> The value has the ISO 639-1 format ("de", "en", etc.) and is limited by the list supported languages.

Do we have a fallback if one sets an unsupported language to use the default?
If not, this must be implemented, if yes we should add this behaviour to readme.md.

**Review by kobergj — APPROVED** [review]

File: services/settings/README.md [hunk 7]
```diff
@@ -71,3 +71,7 @@ Services can set or query ocis *setting values* of a user from settings bundles.
 ## Service Accounts
 
 The settings service needs to know the ID's of service accounts but it doesn't need their secrets. Currently only one service account can be configured which has the admin role. This can be set with the `SETTINGS_SERVICE_ACCOUNT_ID_ADMIN` envvar, but it will also pick up the global `OCIS_SERVICE_ACCOUNT_ID` envvar. Also see the 'auth-service' service description for additional details.
+
+## Default language
```
Comment (mmattel) [inline services/settings/README.md:75]:
```suggestion
## Default Language
```

File: services/settings/README.md [hunk 8]
```diff
@@ -71,3 +71,7 @@ Services can set or query ocis *setting values* of a user from settings bundles.
 ## Service Accounts
 
 The settings service needs to know the ID's of service accounts but it doesn't need their secrets. Currently only one service account can be configured which has the admin role. This can be set with the `SETTINGS_SERVICE_ACCOUNT_ID_ADMIN` envvar, but it will also pick up the global `OCIS_SERVICE_ACCOUNT_ID` envvar. Also see the 'auth-service' service description for additional details.
+
+## Default Language
+
+The default language can be defined via SETTINGS_DEFAULT_LANGUAGE environment variable. If this variable is not defined, English will be used as default. The value has the ISO 639-1 format ("de", "en", etc.) and is limited by the list supported languages. This setting can be used to set the default language for invitation emails.
```
Comment (mmattel) [inline services/settings/README.md:77]:
```suggestion
The default language can be defined via SETTINGS_DEFAULT_LANGUAGE environment variable. If this variable is not defined, English will be used as default. The value has the ISO 639-1 format ("de", "en", etc.) and is limited by the list supported languages. This setting can be used to set the default language for invitation emails.

Important developer note: the list of supported languages is at the moment not easy defineable, as it is the minimum intersection of languages shown in the WebUI and languages defined in the ocis code for the use of notifications. Even more, not all languages where there are translations available on transifex, are available in the WebUI respectively for ocis notifications, and the translation rate for existing languages is partially not that high. You will see therefore quite often English default strings though a supported language may exist and was selected.
```

**Review by mmattel — APPROVED** [review]
Looks good from a docs POV

**Pull request merged**